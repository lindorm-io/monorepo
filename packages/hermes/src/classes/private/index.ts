export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
