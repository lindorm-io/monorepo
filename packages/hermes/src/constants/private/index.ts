export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
