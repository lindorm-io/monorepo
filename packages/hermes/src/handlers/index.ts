export * from "./HermesAggregateCommandHandler";
export * from "./HermesAggregateErrorHandler";
export * from "./HermesAggregateEventHandler";
export * from "./HermesSagaErrorHandler";
export * from "./HermesSagaEventHandler";
export * from "./HermesSagaIdHandler";
export * from "./HermesSagaTimeoutHandler";
export * from "./HermesViewErrorHandler";
export * from "./HermesViewEventHandler";
export * from "./HermesViewIdHandler";
export * from "./HermesViewQueryHandler";
