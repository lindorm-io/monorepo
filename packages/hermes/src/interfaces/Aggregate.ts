import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
