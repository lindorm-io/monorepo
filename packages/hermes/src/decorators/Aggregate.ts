import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
