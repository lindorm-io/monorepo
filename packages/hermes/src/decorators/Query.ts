import { Constructor } from "@lindorm/types";
import { globalHermesMetadata } from "../utils/private";

export function Query(): ClassDecorator {
  return function (target) {
    globalHermesMetadata.addQuery({
      name: target.name,
      target: target as unknown as Constructor,
    });
  };
}
