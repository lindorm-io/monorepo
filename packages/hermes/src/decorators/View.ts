/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { isArray } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { ViewDecoratorOptions, ViewStoreSource } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function View(
  AggregateClass: Constructor | Array<Constructor>,
  source: ViewStoreSource,
  options: ViewDecoratorOptions = {},
): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addView({
      aggregates: isArray(AggregateClass) ? AggregateClass : [AggregateClass],
      name: options.name || name,
      namespace: options.namespace || null,
      source,
      target: target as unknown as Constructor,
    });
  };
}
