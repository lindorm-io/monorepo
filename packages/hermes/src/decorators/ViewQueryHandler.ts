import { Constructor, Dict } from "@lindorm/types";
import { ViewQueryHandlerDescriptor } from "../types";
import { globalHermesMetadata } from "../utils/private";

export function ViewQueryHandler<C extends Constructor, S extends Dict>(
  QueryClass: C,
): ViewQueryHandlerDescriptor<C, S> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: null,
      decorator: "ViewQueryHandler",
      encryption: false,
      handler: descriptor.value,
      key: key.toString(),
      schema: null,
      target: target.constructor as Constructor,
      trigger: QueryClass,
    });
  };
}
