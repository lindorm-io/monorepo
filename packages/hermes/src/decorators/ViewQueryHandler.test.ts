import { Dict } from "@lindorm/types";
import { ViewQueryCtx } from "../types/handlers/view-query-handler";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { Query } from "./Query";
import { View } from "./View";
import { ViewQueryHandler } from "./ViewQueryHandler";

describe("ViewQueryHandler Decorator", () => {
  @Aggregate()
  class TestViewAggregate {}

  test("should add metadata", () => {
    @Query()
    class TestViewQueryHandlerQuery {}

    @View(TestViewAggregate, "mongo")
    class TestViewQueryHandlerView {
      @ViewQueryHandler(TestViewQueryHandlerQuery)
      public async onTestViewQueryHandlerQuery(
        ctx: ViewQueryCtx<TestViewQueryHandlerQuery, Dict>,
      ) {}
    }

    expect(globalHermesMetadata.getView(TestViewQueryHandlerView)).toMatchSnapshot();
  });
});
