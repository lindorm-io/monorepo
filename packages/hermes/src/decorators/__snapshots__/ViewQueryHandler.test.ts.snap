// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewQueryHandler Decorator should add metadata 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [
    {
      "conditions": null,
      "decorator": "ViewQueryHandler",
      "encryption": false,
      "handler": [Function],
      "key": "onTestViewQueryHandlerQuery",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_view_query_handler_view",
  "namespace": null,
  "source": "mongo",
  "target": [Function],
}
`;
